src/vaults/Distributor.ts

```diff
@@ -0,0 +1,228 @@
+import { TransactionEvents, UserConfig } from "./types";
+import { EventsFetcherInterface } from "./VaultEventsFetcher";
+import { BigNumber, constants } from "ethers";
+import { WadRayMath } from "@morpho-labs/ethers-utils/lib/maths";
+import { DepositEvent, TransferEvent, WithdrawEvent } from "./contracts/ERC4626";
+import { formatUnits } from "ethers/lib/utils";
+import { computeMerkleTree } from "../utils";
+import { ProofsFetcherInterface } from "./ProofsFetcher";
+import { pow10 } from "@morpho-labs/ethers-utils/lib/utils";
+import { EpochConfig } from "../ages";
+
+export enum VaultEventType {
+  Deposit = "DEPOSIT",
+  Withdraw = "WITHDRAW",
+  Transfer = "TRANSFER",
+}
+
+type MerkleTree = ReturnType<typeof computeMerkleTree>;
+
+export default class Distributor {
+  static readonly SCALING_FACTOR = pow10(36);
+  public readonly vaultAddress: string;
+
+  private _marketIndex = constants.Zero; // The index saving morpho accrued for Vault users
+  private _lastTimestamp = constants.Zero; // The timestamp of the last marketIndex update
+  private _totalSupply = constants.Zero; // The sum of the shares of all users
+  private _morphoAccumulatedFromMainDistribution = constants.Zero; // The accumulated MORPHO claimable by the vault
+
+  private _usersConfigs: Record<string, UserConfig | undefined> = {}; // Mapping containing the user related information
+
+  constructor(
+    vaultAddress: string,
+    private readonly eventsFetcher: EventsFetcherInterface,
+    private readonly proofsFetcher: ProofsFetcherInterface
+  ) {
+    this.vaultAddress = vaultAddress.toLowerCase();
+  }
+
+  async distributeMorpho(epochToId?: string) {
+    this._clean();
+    const epochsProofs = await this.proofsFetcher.fetchProofs(this.vaultAddress, epochToId);
+    if (!epochsProofs?.length)
+      throw Error(`No MORPHO distributed for the vault ${this.vaultAddress} in epoch ${epochToId}`);
```

Comment:
We can safely assume `epochsProofs` isn't empty here, as `fetchProofs` throws an error already

Reply:
The only uncovered edge-case is when `proofs` fetched is empty:
- either we want to cover it with an error, in which case I recommend throwing it inside `fetchProofs`
- or we don't want to throw an error, in which case we should just return?

Reply:
Since it is not specified that fetchProofs does not return an empty array, I think that we have to keep this check

Reply:
It's not specified, but it's part of the current implementation ; my point being that it could be specified/expected from `fetchProofs`
also, the error would look like "in epoch undefined` in the case `epochToId === undefined`

My generic point is more that if we throw errors, we should throw them in a single place (whereas here we are throwing an error which is supposed to be thrown from a lower level too, so it's redundant)
Either we don't throw the error in `fetchProofs`, or we throw it only in `fetchProofs` (otherwise we've got redundant code)

Reply:
I agree with you on all points, but to me it is better to remove the check from fetchProofs and let the one inside of the distributor